packages/block-editor/src/components/grid-interactivity/grid-item-pin-toolbar-item.js

```diff
@@ -0,0 +1,66 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { ToolbarGroup, ToolbarButton } from '@wordpress/components';
+import { pin as pinIcon } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import BlockControls from '../block-controls';
+import { __unstableUseBlockElement as useBlockElement } from '../block-list/use-block-props/use-block-refs';
+import { getComputedCSS, getGridTracks, getClosestTrack } from './utils';
+
+export function GridItemPinToolbarItem( { clientId, layout, onChange } ) {
+	const blockElement = useBlockElement( clientId );
+	if ( ! blockElement ) {
+		return null;
+	}
+
+	const isPinned = !! layout?.columnStart && !! layout?.rowStart;
+
+	function unpinBlock() {
+		onChange( {
+			columnStart: undefined,
+			rowStart: undefined,
+		} );
+	}
+
+	function pinBlock() {
+		const gridElement = blockElement.parentElement;
+		const columnGap = parseFloat(
+			getComputedCSS( gridElement, 'column-gap' )
+		);
+		const rowGap = parseFloat( getComputedCSS( gridElement, 'row-gap' ) );
+		const gridColumnTracks = getGridTracks(
+			getComputedCSS( gridElement, 'grid-template-columns' ),
+			columnGap
+		);
+		const gridRowTracks = getGridTracks(
+			getComputedCSS( gridElement, 'grid-template-rows' ),
+			rowGap
+		);
+		const columnStart =
+			getClosestTrack( gridColumnTracks, blockElement.offsetLeft ) + 1;
+		const rowStart =
+			getClosestTrack( gridRowTracks, blockElement.offsetTop ) + 1;
+		onChange( {
+			columnStart,
+			rowStart,
+		} );
+	}
+
+	return (
+		<BlockControls group="parent">
+			<ToolbarGroup>
+				<ToolbarButton
+					icon={ pinIcon }
+					label={ __( 'Pin to grid' ) }
```

Comment:
Could we check the state and make this text "Pinned to grid" when it's enabled? Otherwise it's confusing to work out which state it's actually in in text labels mode:

<img width="107" alt="Screenshot 2024-03-15 at 3 16 13 PM" src="https://github.com/WordPress/gutenberg/assets/8096000/55d8d1e4-203f-4b8a-a314-46e1eae6d5c5">